packages/tokenbank_v1/contracts/modules/transfers/QuotaTransfers.sol

```diff
@@ -16,11 +16,259 @@
+            approveInternal(wallet, token, to, allowance);
+            return;
+        }
+
+        revert("OUT_OF_QUOTA");
+    }
+
+    function callContract(
+        address            wallet,
+        address            to,
+        uint               amount,
+        bytes     calldata data,
+        bool               enablePending
+        )
+        external
+        nonReentrant
+        onlyWhenWalletUnlocked(wallet)
+        returns (bytes32 pendingTxId)
+    {
+        bytes32 txid = keccak256(
+            abi.encodePacked(
+                "__CALL_CONTRACT__",
+                wallet,
+                to,
+                amount,
+                keccak256(data)
+            )
+        );
+
+        bool foundPendingTx = authorizeWalletOwnerAndPendingTx(wallet, txid);
+
+        if (whitelistStore.isWhitelisted(wallet, to)) {
+            callContractInternal(wallet, to, amount, data);
+            return bytes32(0);
+        }
+
+        uint valueInCNY = priceProvider.getValueInCNY(address(0), amount);
+        if (quotaStore.checkAndAddToSpent(wallet, valueInCNY)) {
+            callContractInternal(wallet, to, amount, data);
+            return bytes32(0);
+        }
+
+        if (!foundPendingTx && enablePending) {
+            pendingTxId = txid;
+            createPendingTx(wallet, pendingTxId);
+        }
+    }
+
+    function approveThenCallContract(
+        address            wallet,
+        address            token,
+        address            to,
+        uint               amount,
+        bytes     calldata data
+        )
+        external
+        nonReentrant
+        onlyFromWalletOwner(wallet)
+        onlyWhenWalletUnlocked(wallet)
+        notWalletOrItsModule(wallet, to)
+    {
+        if (whitelistStore.isWhitelisted(wallet, to)) {
+            approveInternal(wallet, token, to, amount);
+            callContractInternal(wallet, to, 0, data);
+            return;
+        }
+
+        uint allowance = ERC20(token).allowance(wallet, to);
+        if (allowance >= amount) {
+            approveInternal(wallet, token, to, amount);
+            callContractInternal(wallet, to, 0, data);
+            return;
+        }
+
+        allowance -= amount;
+        uint valueInCNY = priceProvider.getValueInCNY(token, allowance);
+
+        if (quotaStore.checkAndAddToSpent(wallet, valueInCNY)) {
+            approveInternal(wallet, token, to, allowance);
+            callContractInternal(wallet, to, 0, data);
+            return;
+        }
+
+        revert("OUT_OF_QUOTA");
+    }
+
+    function authorizeWalletOwnerAndPendingTx(
+        address wallet,
+        bytes32 pendingTxId
+        )
+        private
+        returns (bool foundPendingTx)
+    {
+        if (msg.sender != Wallet(wallet).owner()) {
+            if (isPendingTxValid(wallet, pendingTxId)) {
+                foundPendingTx = true;
+                emit PendingTxExecuted(wallet, pendingTxId, now);
```

Comment:
Pending transactions aren't marked as executed, so it's valid that pending transactions can be executed an unlimited amount of times within `pendingExpiry` seconds of creating the pending transaction?

Reply:
fixed.